### Business Analytics - Assignment 1 {-}

Umut Demirhan
46739106


---

**Assignment Points**: 100  
**Due Date**: Friday Week 4 (17 March) @ 11.59pm  
**Submission**: Submit your file using the submission link on iLearn


- Put **all your work** into this file;
- Failure to supply solutions in the cells provided below each question will result in a loss of 30 points;
- Follow all instructions closely and **do not** print your variables to the screen unless explicitly asked to do so;
    - Comment out print statements where necessary and make sure that your submitted notebook does not have the output of previously executed print statements;
    - 10 marks will be deducted for every redundant print statement not explicitly asked for.



### About the Assignment

Credit score cards are used as a risk control method in the financial industry. Personal information submitted by credit card applicants are used to predict the probability of future defaults. The bank employs such data to decide whether to issue a credit card to the applicant or not.




| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number      |         |
| Income   | Annual income  |  |
| Gender   | Applicant's Gender   | Male = 0, Female = 1  |
| Car | Car Ownership | Yes = 1, No = 0 | 
| Children | Number of Children | |
| Real Estate | Real Estate Ownership | Yes = 1, No = 0 
| Days Since Birth | No. of Days | Count backwards from current day (0), -1 means yesterday
| Days Employed | No. of Days | Count backwards from current day(0). If positive, it means the person is currently unemployed.
| Payment Default | Whether a client has overdue credit card payments | Yes = 1, No = 0



---


### Problem 1 - (50 points) {-}


**Question 1** 

- Import the `assignment_data.xlsx` file from `data` folder into a pandas DataFrame named `df`; 
- Delete duplicate rows from `df` according to `ID`;
- Delete the `ID` column.
- How many rows are left in `df`?

(10 points)

In [1]:
# Import the data
import pandas as pd
df = pd.read_excel('data/assignment_data.xlsx')

# Deleting duplicate rows from df according to ID
df.drop_duplicates(subset="ID", inplace=True)

#Deleting the ID column
df.drop(columns=['ID'], inplace=True)

#Number of rows lleft in df
df.shape

(5976, 8)

5976 rows are left after dublicates are removed.


---

**Question 2**

- Reset the index in `df` using an appropriate function from `pandas` so that the new index corresponds to the number of rows (make sure to delete the old index). 
- How many positive values of `Days Employed` are there?
- Replace the positive values of `Days Employed` with 0 (zero) in `df`

(10 points)

In [2]:
#Reseting the index in df
df.reset_index(inplace= True,drop=True)

#Number of positive values of Days Employed
(df['Days Employed'] > 0).sum()

#Replacing the positive values of Days Employed with 0  in df
df.loc[df['Days Employed']>0,'Days Employed']=0

* There were 967 positive values of Days Employed

---
**Question 3**

Create two new variables in `df` named 

1. `Age`;
2. `Years in Employment`,

which measure age and employment length in **years** (decimal numbers) from `Days Since Birth` and `Days Employed` by applying approapriate transformations on these variables. 

Delete the original variables `Days Since Birth` and `Days Employed`.

(5 points)


In [3]:
#Creating new variable 'Age' and deleting column 'Days Since Birth'
df['Age'] = -df['Days Since Birth']/365
del df['Days Since Birth']

#Creating new variable 'Years in Employment' and deleting column 'Days Employed'
df['Years in Employment'] = -df['Days Employed']/365\
del df['Days Employed']

---
**Question 4**

- Create a **one**-dimensional NumPy array named `y` by exporting the first 5,000 observations of `Payment_Default`. (Hint: see `ravel()` function)
- Create a NumPy array named `X` by exporting the first 5,000 observations of the following columns `Gender`, `Car`, `Real Estate`, `Children`, `Income`, `Age`, `Years in Employment`.
 
(10 points)


In [28]:
import numpy as np
#Creating a one-dimensional NumPy array named y by exporting the first 5,000 observations of 'Payment_Default'
y = np.array(df.iloc[:5000,5:6])
y = np.ravel(y)

#Creat a NumPy array named X by exporting the first 5,000 observations of all columns except 'Payment_Default'
df1 = df.drop('Payment Default', axis=1)
X = np.array(df1.iloc[:5000])

---

**Question 5** 

- Use an appropriate `scikit-learn` library we learned in class to create the following NumPy arrays: `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 70% training and 30% test datasets. 
- Set `random_state` to 0 and stratify subsamples so that train and test datasets have roughly equal proportions of the target's class labels. 

(5 points) 

In [5]:
# Splitting the data into 70% training and 30% test datasets. 
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 0, stratify = y)

---

**Question 6**

- Create new variables by using an appropriate `scikit-learn` library we learned in class to standardize the features from the training and test datasets to mean zero and variance one. Name the new variables by appending '_scaled' to the original variable names.


(10 points)   

In [6]:
# Creating new variables by standardizing the features from the training and test datasets
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()

sc.fit(X_train)

X_train_scaled = sc.transform(X_train)
X_test_scaled = sc.transform(X_test)

---

## Problem 2 - (20 Points) {-}

**Question 7**

Fit the following two classifiers to the transformed training dataset using `scikit-learn` libraries.

- Perceptron - name your instance `pc` set `random_state=1`
- Logistic Regression - name your instance `lr` set `random_state=1`

When initializing instances of the above classifiers only set the parameters referenced above and nothing else.

(20 points)

In [7]:
#Perceptron
from sklearn.linear_model import Perceptron
pc = Perceptron(random_state=1)
pc.fit(X_train_scaled, y_train)

#Logistic Regression
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(random_state=1)
lr.fit(X_train_scaled, y_train)

LogisticRegression(random_state=1)

---
## Problem 3 - (30 points) {-}


**Question 8**

- Using a method built into each of the two classifiers compute their prediction accuracies on the training data;
- Store the accuracy values into variables named according to the following pattern: `classifier_name_accuracy_train`, e.g. you should have `lr_accuracy_train`; 
- Print the two accuracy **variables** along with their brief descriptions.

(10 points)

In [25]:
from sklearn.metrics import accuracy_score

#Accuracy of Perceptron on the training data
y_train_pred_pc = pc.predict(X_train_scaled)
Perceptron_accuracy_train = accuracy_score(y_train, y_train_pred_pc)
print(f'Accuracy of Perceptron = {Perceptron_accuracy_train:.3f}')

# Accuracy of Logistic Regression  on the training data
y_train_pred_lr = lr.predict(X_train_scaled)
lr_accuracy_train = accuracy_score(y_train, y_train_pred_lr)
print(f'Accuracy of Logistic Regression = {lr_accuracy_train:.3f}')

Accuracy of Perceptron = 0.496
Accuracy of Logistic Regression = 0.558


---

**Question 9** 

- Using a method built into each of the above classifiers compute their prediction accuracy for the test dataset
- Store the accuracy values into variables named according to the following pattern: `classifier_name_accuracy_test`, e.g. you should have `lr_accuracy_test`. 
- Print the two accuracy **variables** along with brief descriptions.

(10 points)

In [27]:
#Accuracy of Perceptron on the test data
y_test_pred_pc = pc.predict(X_test_scaled)
Perceptron_accuracy_test = accuracy_score(y_test, y_test_pred_pc)
print(f'Accuracy of Perceptron = {Perceptron_accuracy_test:.3f}')

# Accuracy of Logistic Regression  on the test data
y_test_pred_lr = lr.predict(X_test_scaled)
lr_accuracy_test = accuracy_score(y_test, y_test_pred_lr)
print(f'Accuracy of Logistic Regression = {lr_accuracy_test:.3f}')

Accuracy of Perceptron = 0.495
Accuracy of Logistic Regression = 0.507


---

**Question 10** 

Using nicely formated text in Markdown comment on the accuracies computed in Questions 8 & 9 making sure you address:
- training and test set datasets; 
- Perceptrion and Logistic Regression models. 

Are the results as expected, and why or why not? (Hint: You are not expected to comment on why a particular model is better.) 

(10 marks)


**Conclusion**


- training and test datasets; the datasets splited into training(%70) and test(%30) datasets so that algorithms can be trained on train dataset and tested on testing dataset.


- Perceptrion and Logistic Regression models: Both algoriths predicted the target variable quite poorly, but logistic regression model performed relatively better as expected due to, I think, some booliean variables. 

- Additionally, as expected accuracy scores on testing data are not better than training data because We use the training data to train the dataset so the test dataset was totaly new dataset to the algorithm to predict.
